### PR TITLE
Added dygraph quantization.

### PR DESCRIPTION
**Pull request by huangxu96**
This Pull Request implemented  an interface of quant aware training in paddleslim. Users can simply add one line code to use quant aware training in paddleslim.

For example:

```python
from paddleslim.quant import quant_aware
quant_model = quant_aware(model)
 
```
The dygraph quant aware use the same interface with static graph quant aware.
We use singledispatch to do it.

### CODE REVIEW

File: paddleslim/quant/dy_quanter.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+def dy_quant_aware(model,
+                config=None,
+                scope=None,
+                for_test=False,
+                weight_quantize_func=None,
+                act_quantize_func=None,
+                weight_preprocess_func=None,
+                act_preprocess_func=None,
+                optimizer_func=None,
+                executor=None):
```
Comment (qingqing01) [inline paddleslim/quant/dy_quanter.py]: 1. Need to add comments and usage
2. “dy_quant_aware”的名字需要线下讨论下~

File: paddleslim/quant/quanter.py [hunk 2]
```diff
@@ -299,6 +304,61 @@ def quant_aware(program,
         quant_program = paddle.static.CompiledProgram(main_graph.graph)
     return quant_program
 
+@quant_aware.register(paddle.fluid.Layer)
```
Comment (baiyfbupt) [inline paddleslim/quant/quanter.py]: 代码适配Paddle2.0，尽量避免用 fluid

File: paddleslim/quant/quanter.py [hunk 3]
```diff
@@ -299,6 +304,61 @@ def quant_aware(program,
...
+                be quantized directly.
+                Default is None.
+        act_preprocess_func(function): Function that defines how to preprocess activation before quantization. Using this
+                can quickly test if user's preprocess method works or not. The function's input
+                is non-quantized activation and function returns processed activation to be quantized. If None, the activation will
+                be quantized directly.
+                Default is None.
+
+    Returns:
+    model(fluid.Layer) | fluid.layer: model with fake quantized layers
```
Comment (baiyfbupt) [inline paddleslim/quant/quanter.py]: 同上

File: tests/test_dygraph_quant_aware.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+from paddle.fluid.dygraph.nn import Pool2D
+from paddle.fluid.dygraph.nn import Linear
+from paddle.fluid.log_helper import get_logger
+
+from paddleslim.quant import quant_aware
+
+_logger = get_logger(
+    __name__, logging.INFO, fmt='%(asctime)s-%(levelname)s: %(message)s')
+
+class ImperativeLenet(fluid.dygraph.Layer):
```
Comment (baiyfbupt) [inline tests/test_dygraph_quant_aware.py]: 同 https://github.com/PaddlePaddle/PaddleSlim/pull/435/files#r517746032

File: paddleslim/quant/quanter.py (hunk 2 above)
Comment (huangxu96) [reply paddleslim/quant/quanter.py]: Done.

File: paddleslim/quant/quanter.py (hunk 3 above)
Comment (huangxu96) [reply paddleslim/quant/quanter.py]: Done.

**Comment by baiyfbupt** [comment]
请在`PaddleSlim/requirements.txt` 中添加一下 singledispatch

File: tests/test_dygraph_quant_aware.py (hunk 4 above)
Comment (huangxu96) [reply tests/test_dygraph_quant_aware.py]: Done.

**Comment by huangxu96** [comment]
> 请在`PaddleSlim/requirements.txt` 中添加一下 singledispatch

**Review by baiyfbupt — APPROVED** [review]
LGTM

File: tests/test_dygraph_quant_aware.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+import logging
+import paddle
+import paddle.nn as nn
+import paddle.fluid as fluid
+from paddle.fluid.optimizer import AdamOptimizer
+from paddle.fluid.dygraph.container import Sequential
+from paddle.fluid.dygraph.nn import Conv2D
+from paddle.fluid.dygraph.nn import Pool2D
+from paddle.fluid.dygraph.nn import Linear
+from paddle.fluid.log_helper import get_logger
```
Comment (qingqing01) [inline tests/test_dygraph_quant_aware.py:28]: 1. 单测还得修改下，采用paddle.nn等

**Pull request merged**